state/mongodb/mongodb.go

```diff
@@ -192,20 +205,70 @@ func (m *MongoDB) setInternal(ctx context.Context, req *state.SetRequest) error
 	if req.ETag != nil {
 		filter[etag] = *req.ETag
 	} else if req.Options.Concurrency == state.FirstWrite {
-		filter[etag] = uuid.NewString()
+		uuid, err := uuid.NewRandom()
+		if err != nil {
+			return err
+		}
+		filter[etag] = uuid.String()
+	}
+
+	reqTTL, err := stateutils.ParseTTL(req.Metadata)
+	if err != nil {
+		return fmt.Errorf("failed to parse TTL: %w", err)
 	}
 
-	update := bson.M{"$set": bson.M{id: req.Key, value: v, etag: uuid.NewString()}}
-	_, err := m.collection.UpdateOne(ctx, filter, update, options.Update().SetUpsert(true))
+	etagV, err := uuid.NewRandom()
+	if err != nil {
+		return err
+	}
 
-	return err
+	update := mongo.Pipeline{
```

Comment:
A bit of a nit, but doing this initialization + an `append` later will cause Go to allocate the slice with capacity 1 and then will need to re-allocate when you do an `append`. Example: https://go.dev/play/p/uryppFGd3li

Consider doing something like (I haven't tested it)

```go
	update := make(mongo.Pipeline, 2)
	update[0] = primitive.D{{"$set", bson.D{
			{id, req.Key},
			{value, v},
			{etag, etagV.String()},
		}}}

	if reqTTL != nil {
		update[1] = primitive.D{{"$addFields", bson.D{
			{daprttl, bson.D{
				{"$dateAdd",
					bson.D{{"startDate", "$$NOW"}, {"unit", "second"}, {"amount", *reqTTL}},
				},
			}},
		}}}
	} else {
		update[1] = primitive.D{{"$addFields", bson.D{
			{daprttl, nil},
		}}}
	}
```

Reply:
Thanks, done